velox/functions/prestosql/window/LeadLag.cpp

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+
+namespace facebook::velox::window::prestosql {
+
+namespace {
+
+template <bool isLag>
+class LeadLagFunction : public exec::WindowFunction {
+ public:
+  explicit LeadLagFunction(
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool, nullptr) {
+    valueIndex_ = args[0].index.value();
+
+    initializeOffset(args);
+    initializeDefaultValue(args);
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+    partitionOffset_ = 0;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& /*frameEnds*/,
+      const SelectivityVector& validRows,
```

Comment:
@mbasmanova : One high level question. For empty frames, is the result NULL or default value ? Since validRows is not handled in code here it seems like there are some assumptions about it that are not explicit.

Reply:
These functions do not consider frames at all. They operate on whole partitions. Hence, it doesn't matter whether frame is empty or not.

Reply:
Interesting. Ranking functions still return null values for empty frames even if they don't honor frames. Maybe add that as a comment and comment out the validRows parameter in this function declaration to make that obvious.

Thanks.

Reply:
> Ranking functions still return null values for empty frames even if they don't honor frames. 

I wonder if this is a bug. Here is what I see in Presto.

```
presto:di> select x, row_number() over (rows between 100 preceding and 90 preceding) from unnest(array[1,2,3,4,5]) as t(x);
 x | _col1
---+-------
 1 |     1
 2 |     2
 3 |     3
 4 |     4
 5 |     5
(5 rows)

presto:di> select x, lag(x) over (rows between 100 preceding and 90 preceding) from unnest(array[1,2,3,4,5]) as t(x);
 x | _col1
---+-------
 1 | NULL
 2 |     1
 3 |     2
 4 |     3
 5 |     4
(5 rows)

presto:di> select x, first_value(x) over (rows between 100 preceding and 90 preceding) from unnest(array[1,2,3,4,5]) as t(x);
 x | _col1
---+-------
 1 | NULL
 2 | NULL
 3 | NULL
 4 | NULL
 5 | NULL
(5 rows)
```

Reply:
That is possible. But that behavior is consistent with DuckDB. Let me research this further.

Reply:
Aditi confirmed that this is a bug: #5154